compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -624,12 +624,15 @@ impl<'tcx> Cx<'tcx> {
                             }
                         }
                         hir::InlineAsmOperand::Const { ref anon_const } => {
-                            let value = mir::Const::identity_unevaluated(
-                                tcx,
-                                anon_const.def_id.to_def_id(),
-                            )
-                            .instantiate_identity()
-                            .normalize(tcx, self.param_env);
+                            let def_id = anon_const.def_id.to_def_id();
+                            let const_type = tcx.type_of(def_id).skip_binder();
+                            let value = if let ty::Error(terr) = const_type.kind() {
+                                mir::Const::Ty(const_type, ty::Const::new_error(tcx, *terr))
+                            } else {
+                                mir::Const::identity_unevaluated(tcx, def_id)
+                                    .instantiate_identity()
+                                    .normalize(tcx, self.param_env)
+                            };
```

Comment:
looking at this I realized that there's something else going on here. The constant starts out as unevaluated and we are able to evaluate it successfully even though it uses a type error as the expected type. We tend to bail in const eval if typeck encountered a type error: https://github.com/rust-lang/rust/blob/9b82580c7347f800c2550e6719e4218a60a80b28/compiler/rustc_const_eval/src/interpret/eval_context.rs#L255-L266

So imo the actual issue is that we don't taint the MIR body even though its expected return type is a type error. We should eagerly taint the infcx using `set_tainted_by_errors` in `typeck_with_fallback` if the expected type is an error. cc @oli-obk as you may be aware of a more general approach to taint the MIR body here

Reply:
Any types written to the typecktables will taint the typeck result. Unfortunately this doesn't include the signature, as that is derived separately sometimes and may independently have errors. Not sure yet what the best way to solve this is, but finding any way and adding a FIXME and issue with the tainting project set is fine

Reply:
followup questions

you talk about the signature and the expected type, but it looks like in the code of `typeck_with_fallback` only the `if` branch explicitly uses the signature, and the else branch has a variable called `expected_type`. 

https://github.com/rust-lang/rust/blob/d6c8169c186ab16a3404cd0d0866674018e8a19e/compiler/rustc_hir_typeck/src/lib.rs#L152-L183

So, in which of the two branches should what be tainted? E.g. is this in the else branch enough?

```rust
// FIXME: link to tainting issue
if let ty::Error(terr) = expected_type.kind() {
    fcx.infcx.set_tainted_by_errors(*terr);
}
```

Or is the `fn_sig` in the if branch really the expected type (it's the type of the definition, which gets checked against the type at the usage site)?

Also, there is no project or label named "tainting" at the moment? nor "taint", does it have a different name?

Reply:
I don't think there's a proper project for it, correctly tainting the MIR body is necessary to avoid followup ICE when trying to evaluate malformed bodies and this is something @oli-obk has recently been spending some significant effort on.

The `fn_sig` is the expected type when type checking a function. For consts, statics, and enum discriminants the expected type is from the `else` branch. So yes, for this concrete issue only tainting in the else branch is enough. Please use `expected_type.error_reported()` instead of matching on `kind` to also handle e.g. references to `<type error>`

Reply:
There is a github project for it https://github.com/orgs/rust-lang/projects/51

Reply:
this should not be unnecessary?

Reply:
you mean that the logic added in earlier commits of this PR can be removed here? If so, that does not work, and again triggers the original ICE

```
thread 'rustc' panicked at compiler/rustc_borrowck/src/type_check/mod.rs:388:75:
called `Option::unwrap()` on a `None` value
```

Reply:
then we are still not tainting enough :thinking:

Reply:
Ah no, we only error in `type_of`, which is computed by using `typeck`. So `typeck` succeeds with `&whatever` as the return type, and we then only replace it with a type error in `type_of` :thinking:

Reply:
so we should do it in https://github.com/rust-lang/rust/blob/master/compiler/rustc_borrowck/src/universal_regions.rs#L583 instead, or after calling that function :thinking:

Reply:
Well, I'm kind of lost there. There no longer is a type to check for "is it an error" at the call site of that function? and within it, it's not so obvious either in the `DefiningTy::Const(self.mir_def.to_def_id(), args)` case.

Reply:
sorry, we only look at the actual type in `fn compute_inputs_and_output`

Reply:
right, well this still does not taint enough

```rust
            DefiningTy::Const(def_id, _) => {
                // For a constant body, there are no inputs, and one
                // "output" (the type of the constant).
                assert_eq!(self.mir_def.to_def_id(), def_id);
                let ty = tcx.type_of(self.mir_def).instantiate_identity();

                // FIXME(#129952): We probably want a more principled approach here.
                if let Err(terr) = ty.error_reported() {
                    self.infcx.set_tainted_by_errors(terr);
                }

                let ty = indices.fold_to_region_vids(tcx, ty);

                ty::Binder::dummy(tcx.mk_type_list(&[ty]))
            }
```

I checked that this block is reached, but I still get that ICE mentioned earlier.

Reply:
is `set_tainted_by_errors` triggered here?

Reply:
ah, we don't propagate the `tainted_by_errors` of the `infcx` in `MirBorrowckCtxt::emit_errors`.

https://github.com/rust-lang/rust/blob/842d6fc32e3d0d26bb11fbe6a2f6ae2afccc06cb/compiler/rustc_borrowck/src/lib.rs#L2525 this should start out with `self.infcx.tainted_by_errors()'

Reply:
yes, e.g. if I add some debug prints

```rust
    // FIXME(#129952): We probably want a more principled approach here.
    if let Err(terr) = ty.error_reported() {
        dbg!("tainting the thing", def_id, &defining_ty);
        self.infcx.set_tainted_by_errors(terr);
    }
```

I get

```
error: invalid type for `const` operand
  --> /home/folkertdev/rust/rust/tests/ui/asm/const-refs-to-static.rs:12:19
   |
LL | global_asm!("{}", const addr_of!(FOO));
   |                   ^^^^^^-------------
   |                         |
   |                         is a `*const u8`
   |
   = help: `const` operands must be of an integer type

[compiler/rustc_borrowck/src/universal_regions.rs:802:21] "tainting the thing" = "tainting the thing"
[compiler/rustc_borrowck/src/universal_regions.rs:802:21] def_id = DefId(0:11 ~ const_refs_to_static[9a35]::{global_asm#0}::{constant#0})
[compiler/rustc_borrowck/src/universal_regions.rs:802:21] &defining_ty = Const(
    DefId(0:11 ~ const_refs_to_static[9a35]::{global_asm#0}::{constant#0}),
    [],
)
error: invalid type for `const` operand
  --> /home/folkertdev/rust/rust/tests/ui/asm/const-refs-to-static.rs:17:25
   |
LL |     unsafe { asm!("{}", const addr_of!(FOO)) };
   |                         ^^^^^^-------------
   |                               |
   |                               is a `*const u8`
   |
   = help: `const` operands must be of an integer type

[compiler/rustc_borrowck/src/universal_regions.rs:802:21] "tainting the thing" = "tainting the thing"
[compiler/rustc_borrowck/src/universal_regions.rs:802:21] def_id = DefId(0:12 ~ const_refs_to_static[9a35]::inline::{constant#0})
[compiler/rustc_borrowck/src/universal_regions.rs:802:21] &defining_ty = Const(
    DefId(0:12 ~ const_refs_to_static[9a35]::inline::{constant#0}),
    [],
)
thread 'rustc' panicked at compiler/rustc_borrowck/src/type_check/mod.rs:388:75:
called `Option::unwrap()` on a `None` value
```

Reply:
cool that works. Is the most recent location for the tain still the correct one then? or should it happen at one of the earlier places?

also this does now fix two crashes from `tests/crashes`